File: src/frontend/src/services/room.ts
import Alpine from 'alpinejs'

export interface RoomType {
  id: string
  revealed: boolean
  parcipants: {
    id: string
    name: string
    vote?: string
  }[]
}

export const joinRoom = async ({ roomId, userName }): Promise<string> => {
  try {
    const url = `/api/room/${roomId}/join`
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ userName }),
    })
    if (!response.ok) {
      throw new Error(`Response status: ${response.status}`)
    }

    const json = await response.json()
    return json.user
  } catch (error) {
    console.error(error.message)
  }
}

let serverSocket: WebSocket

export const connectRoom = ({ roomId, userId }) => {
  Alpine.store('room').id = roomId
  Alpine.store('user').id = userId

  if (!roomId || !userId) {
    return
  }

  const prot = window.location.protocol === 'http:' ? 'ws:' : 'wss:'

  const socketUrl = `${prot}//${window.location.host}/api/ws/roomhub/${roomId}/${userId}`
  serverSocket = new WebSocket(socketUrl)

  serverSocket.onmessage = (ev) => {
    const { type, data } = JSON.parse(ev.data)
    console.log('RX', type, data)

    if (type === 'roomUpdate') {
      Alpine.store('room', data)
    }
  }

  serverSocket.onclose = () => {
    console.error('Socket closed unexpectedly')
  }
}

const _send = (type, data) => {
  console.log('TX', type, data)
  serverSocket.send(JSON.stringify({ type, data }))
}

export const vote = (points) => _send('vote', points)
export const unvote = () => _send('unvote', {})
export const resetVotes = () => _send('reset', {})
export const reveal = () => _send('reveal', {})
export const unreveal = () => _send('unreveal', {})
export const removeUser = (userId) => _send('removeUser', userId)

export const myVote = () => {
  const userId = Alpine.store('user').id
  return Alpine.store('room')?.participants?.find(({ id }) => id == userId)
    ?.vote
}

export const consensus = () => {
  const parts = Alpine.store('room')?.participants
  const votes = parts?.filter(({ vote }) => vote).map(({ vote }) => vote)
  return votes.length > 1 && votes.every(({ vote }) => vote == votes[0])
}

File: src/backend/pointhub/services/room.py

from pointhub.utils import create_id

from . import cache

ROOM_EXPIRATION_S = 60*60

def _add_participant(room, name):
    participant = {'id': create_id(), 'name': name}
    room['participants'].append(participant)
    return participant

def _get_room(c, room_id):
    obj = c.json().get(room_id, '$')
    return obj[0] if obj else None

def _save_room(c, room):
    c.json().set(room['id'], '$', room)
    c.expire(room['id'], time=ROOM_EXPIRATION_S)

def _is_owner(room, user_id):
    return room['participants'][0]['id'] == user_id

def create(owner_name):
    room = {
        'id': create_id(),
        'revealed': False,
        'participants': []
    }
    _add_participant(room, owner_name)

    with cache.connection() as c:
        _save_room(c, room)

    return room

def get(room_id):
    with cache.connection() as c:
        return _get_room(c, room_id)

def join(room_id, user_name):
    with cache.connection() as c:
        room = _get_room(c, room_id)
        if not room:
            return None

        user_id = _add_participant(room, user_name)['id']
        _save_room(c, room)
        return user_id

def vote(room_id, user_id, points):
    with cache.connection() as c:
        room = _get_room(c, room_id)
        if not room:
            return None

        user = next((u for u in room['participants'] if u['id'] == user_id), None)
        if not user:
            return None

        if points is None:
            user.pop('vote', None)
        else:
            user['vote'] = points
        _save_room(c, room)
        return room

def set_reveal(room_id, revealed):
    with cache.connection() as c:
        room = _get_room(c, room_id)
        if not room:
            return None

        room['revealed'] = revealed
        _save_room(c, room)
        return room

def reset_votes(room_id):
    with cache.connection() as c:
        room = _get_room(c, room_id)
        if not room:
            return None

        for user in room['participants']:
            user.pop('vote', None)
        room['revealed'] = False
        _save_room(c, room)
        return room

def remove_user(room_id, req_user_id, del_user_id):
    with cache.connection() as c:
        room = _get_room(c, room_id)
        if not room or not _is_owner(room, req_user_id) or req_user_id == del_user_id:
            return None

        room['participants'] = [p for p in room['participants'] if p['id'] != del_user_id]
        _save_room(c, room)
        return room
